src/i_winmusic.c

```diff
@@ -291,17 +311,195 @@ static void ResetDevice(void)
+                             msg[5] == 0x7F && // Address Low
+                             msg[6] == 0x00)   // Data
+                    {
+                        // XG All Parameter Reset
+                        // 43 <dev> 4C 00 00 7F 00 F7
+                        return true;
+                    }
+                    break;
+            }
+            break;
+
+        case 0x7E: // Universal Non-Real Time
+            switch (msg[2])
+            {
+                case 0x09: // General Midi
+                    if (length == 5 &&
+                       (msg[3] == 0x01 || // GM System On
+                        msg[3] == 0x02 || // GM System Off
+                        msg[3] == 0x03))  // GM2 System On
+                    {
+                        // GM System On/Off, GM2 System On
+                        // 7E <dev> 09 01 F7
+                        // 7E <dev> 09 02 F7
+                        // 7E <dev> 09 03 F7
+                        return true;
+                    }
+                    break;
+            }
+            break;
+    }
+    return false;
+}
+
+static boolean IsMasterVolume(const byte *msg, int length, unsigned int *volume)
+{
+    // General Midi (7F <dev> 04 01 <lsb> <msb> F7)
+    if (length == 7 &&
+        msg[0] == 0x7F && // Universal Real Time
+        msg[2] == 0x04 && // Device Control
+        msg[3] == 0x01)   // Master Volume
+    {
+        *volume = (msg[4] & 0x7F) | ((msg[5] & 0x7F) << 7);
+        return true;
+    }
+
+    // Roland (41 <dev> 42 12 40 00 04 <vol> <sum> F7)
+    if (length == 10 &&
+        msg[0] == 0x41 && // Roland
+        msg[2] == 0x42 && // GS
+        msg[3] == 0x12 && // DT1
+        msg[4] == 0x40 && // Address MSB
+        msg[5] == 0x00 && // Address
+        msg[6] == 0x04)   // Address LSB
+    {
+        *volume = DEFAULT_MASTER_VOLUME * (msg[7] & 0x7F) / 127;
+        return true;
+    }
+
+    // Yamaha (43 <dev> 4C 00 00 04 <vol> F7)
+    if (length == 8 &&
+        msg[0] == 0x43 && // Yamaha
+        msg[2] == 0x4C && // XG
+        msg[3] == 0x00 && // Address High
+        msg[4] == 0x00 && // Address Mid
+        msg[5] == 0x04)   // Address Low
+    {
+        *volume = DEFAULT_MASTER_VOLUME * (msg[6] & 0x7F) / 127;
+        return true;
+    }
+
+    return false;
+}
+
+static void SendSysExMsg(int time, const byte *data, int length)
+{
+    native_event_t native_event;
+
+    if (IsMasterVolume(data, length, &master_volume))
+    {
+        // Found a master volume message in the MIDI file. Take this new
+        // value and scale it by the user's volume slider.
+        UpdateVolume(time);
+        return;
+    }
+
+    // Send the SysEx message.
+    native_event.dwDeltaTime = time;
+    native_event.dwStreamID = 0;
+    native_event.dwEvent = MAKE_EVT(length + sizeof(byte), 0, 0, MEVT_LONGMSG);
+    WriteBuffer((byte *)&native_event, sizeof(native_event_t));
+    WriteBuffer(&(byte){MIDI_EVENT_SYSEX}, sizeof(byte));
```

Comment:
I think the maintainers of Chocolate Doom won't like `&(byte){MIDI_EVENT_SYSEX}`, I suggest this little change:
```diff
diff --git a/src/i_winmusic.c b/src/i_winmusic.c
index a1c6d74..0a2af57 100644
--- a/src/i_winmusic.c
+++ b/src/i_winmusic.c
@@ -476,6 +476,7 @@ static boolean IsMasterVolume(const byte *msg, int length, unsigned int *volume)
 static void SendSysExMsg(int time, const byte *data, int length)
 {
     native_event_t native_event;
+    const char event_type = MIDI_EVENT_SYSEX;
 
     if (IsMasterVolume(data, length, &master_volume))
     {
@@ -490,7 +491,7 @@ static void SendSysExMsg(int time, const byte *data, int length)
     native_event.dwStreamID = 0;
     native_event.dwEvent = MAKE_EVT(length + sizeof(byte), 0, 0, MEVT_LONGMSG);
     WriteBuffer((byte *)&native_event, sizeof(native_event_t));
-    WriteBuffer(&(byte){MIDI_EVENT_SYSEX}, sizeof(byte));
+    WriteBuffer(&event_type, sizeof(byte));
     WriteBuffer(data, length);
     WriteBufferPad();
 
```

Reply:
Fixed, thank you!